pkg/object/function/worker/ingress.go

```diff
@@ -0,0 +1,328 @@
+	b.Filters = append(b.Filters, map[string]interface{}{
+		"kind": proxy.Kind,
+		"name": backendName,
+		"mainPool": &proxy.PoolSpec{
+			Servers:     mainServers,
+			LoadBalance: lb,
+		},
+	})
+
+	return b
+}
+
+// Get gets ingressServer itself as the default backend.
+// egress server will handle the pipeline routing by itself.
+func (ings *ingressServer) Get(name string) (protocol.HTTPHandler, bool) {
+	ings.mutex.RLock()
+	defer ings.mutex.RUnlock()
+	return ings, true
+}
+
+func (ings *ingressServer) httpServerSpec(httpServer *httpserver.Spec) string {
+	ingressHTTPServerFormat := `
+kind: HTTPServer
+name: %s
+http3: %t
+port: %d
+keepAlive: %t
+keepAliveTimeout: %s
+https: %t
+certBase64: %s
+keyBase64: %s
+maxConnections: %d
+rules:
+  - paths:
+    - pathPrefix: /
+      backend: accept-all-placeholder`
+
+	return fmt.Sprintf(ingressHTTPServerFormat,
+		ings.superSpec.Name(), httpServer.HTTP3, httpServer.Port, httpServer.KeepAlive,
+		httpServer.KeepAliveTimeout, httpServer.HTTPS, httpServer.CertBase64,
+		httpServer.KeyBase64, httpServer.MaxConnections)
+}
+
+// Init creates a default ingress HTTPServer.
+func (ings *ingressServer) Init() error {
+	ings.mutex.Lock()
+	defer ings.mutex.Unlock()
+
+	if ings.httpServer != nil {
+		return nil
+	}
+	spec := ings.superSpec.ObjectSpec().(*spec.Admin)
+
+	ings.faasNetworkLayerURL = spec.Knative.NetworkLayerURL
+	ings.faasHostSuffix = spec.Knative.HostSuffix
+	ings.faasNamespace = spec.Knative.Namespace
+
+	yamlConf := ings.httpServerSpec(spec.HTTPServer)
+	logger.Debugf("http svr spec is %s", yamlConf)
+
+	superSpec, err := supervisor.NewSpec(string(yamlConf))
+	if err != nil {
+		logger.Errorf("new spec for %s failed: %v", yamlConf, err)
+		return err
+	}
+
+	var httpServer httpserver.HTTPServer
+	httpServer.Init(superSpec, ings.super)
+	httpServer.InjectMuxMapper(ings)
+	ings.httpServer = &httpServer
+	return nil
+}
+
+// Put puts pipeline named by faas function's name with a requestAdaptor and proxy
+func (ings *ingressServer) Put(funcSpec *spec.Spec) error {
+	builder := newPipelineSpecBuilder(funcSpec.Name)
+	builder.appendReqAdaptor(funcSpec, ings.faasNamespace, ings.faasHostSuffix)
+	builder.appendProxy(ings.faasNetworkLayerURL)
+
+	yamlConfig := builder.yamlConfig()
+	logger.Debugf("pipeline spec is %s", yamlConfig)
+
+	superSpec, err := supervisor.NewSpec(yamlConfig)
+	if err != nil {
+		logger.Errorf("new spec for %s failed: %v", yamlConfig, err)
+		return err
+	}
+	pipeline := &httppipeline.HTTPPipeline{}
+
+	pipeline.Init(superSpec, ings.super)
+	ings.pipelines[funcSpec.Name] = &faasPipeline{pipeline: pipeline, active: false}
+
+	return nil
+}
+
+// Delete deletes one ingress pipeline according to the function's name.
+func (ings *ingressServer) Delete(functionName string) {
+	ings.mutex.Lock()
+	defer ings.mutex.Unlock()
+
+	if p, ok := ings.pipelines[functionName]; ok {
+		p.pipeline.Close()
+		delete(ings.pipelines, functionName)
+	}
```

Comment:
Propose to move some code out from the lock to improve concurrency:

```suggestion
	ings.mutex.Lock()
	p, ok := ings.pipelines[functionName]
	if ok {
		delete(ings.pipelines, functionName)
	}
	ings.mutex.Unlock()
	if ok {
		p.pipeline.Close()
	}
```